src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1730,6 +1730,18 @@ bool checkValid(MethodSymbol candidate, bool isCandidate, DiagnosticBag specific
                         }
                     }
                 }
+                else if (LanguageVersion >= MessageID.IDS_FeatureAsyncMain.RequiredVersion() && taskEntryPoints.Count > 0)
+                {
```

Comment:
This logic doesn't seem correct to me. Please add a test with 2 valid sync `Main`s and 1 valid async `Mains`. I would expect this section to report warnings on all async `Main` methods, and report a warning on the sync `Main` if there is only 1 sync Main. If there is more than 1 sync `Main`, then it will be errored on below.

Reply:
@333fred The `Main` method can either be parameterless or have one string array parameter.
So, shouldn't defining 3 `Main`s report an error `(error CS0111: Type 'type' already defines a member called 'Main' with the same parameter types)` anyway? whether all the three are sync, or all async, or 2 sync and 1 async, or 1 sync and 2 async?

Reply:
```
Expected:
                Diagnostic(ErrorCode.WRN_SyncAndAsyncEntryPoints, "Main")
Actual:
                // (7,22): warning CS8892: Program has more than one entry point defined.
                //   public static void Main(string[] a) { System.Console.WriteLine(2); }
                Diagnostic(ErrorCode.WRN_SyncAndAsyncEntryPoints, "Main").WithLocation(7, 22),
                // (7,22): error CS0017: Program has more than one entry point defined. Compile with /main to specify the type that contains the entry point.
                //   public static void Main(string[] a) { System.Console.WriteLine(2); }
                Diagnostic(ErrorCode.ERR_MultipleEntryPoints, "Main").WithLocation(7, 22)
Diff:
                    Diagnostic(ErrorCode.WRN_SyncAndAsyncEntryPoints, "Main")
++>                 Diagnostic(ErrorCode.ERR_MultipleEntryPoints, "Main").WithLocation(7, 22)
```

The test with 2 sync and 1 async reported the new warning CS8892, and the old error CS0017.

@333fred it should only report the error right?

So, I may check the if condition to be `else if (LanguageVersion >= MessageID.IDS_FeatureAsyncMain.RequiredVersion() && taskEntryPoints.Count == 1 && viableEntryPoints.Count ==1)`, does that seem correct to you?

Reply:
I don't think the `taskEntryPoints.Count == 1` part is correct. It should be any non-zero number of task entry points, and 1 regular entry point.

---
In reply to: [472445461](https://github.com/dotnet/roslyn/pull/46832#discussion_r472445461) [](ancestors = 472445461)

Reply:
Make sure to add a test with multiple async entry points, and one sync entry point.

---
In reply to: [472473091](https://github.com/dotnet/roslyn/pull/46832#discussion_r472473091) [](ancestors = 472473091,472445461)

Reply:
> I don't think the taskEntryPoints.Count == 1 part is correct. It should be any non-zero number of task entry points, and 1 regular entry point.

Shouldn't the case of multiple 2 async main produce an error, not a warning?

Reply:
https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgBgAJVMBWAbjz1QGYT0iBhIgbzyM4cdYF8OudEgDYiAWQCGASygAKUgQDaAXSIBnAJRsSAdhEA6JgHsAtgAcANgFNgVuKmEUi/XC6EZmPAZ3ejJM+UwlVU1tVD0HQ1NLGzsHJxdeIA

Reply:
> 
> 
> https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgBgAJVMBWAbjz1QGYT0iBhIgbzyM4cdYF8OudEgDYiAWQCGASygAKUgQDaAXSIBnAJRsSAdhEA6JgHsAtgAcANgFNgVuKmEUi/XC6EZmPAZ3ejJM+UwlVU1tVD0HQ1NLGzsHJxdeIA

@333fred is that a compiler crash??

Reply:
What I'm getting locally doesn't match that strange stack trace in sharplab:

![image](https://user-images.githubusercontent.com/31348972/90568574-757d6c80-e1ac-11ea-8055-719069cbdce1.png)

Am I missing something?

Reply:
Looks like Fred's sharplab link just hits a limitation of the website, it doesn't allow you to execute code involving Task. The link does however demonstrate that a main method was chosen successfully, because if we hadn't we would have seen the diagnostics instead of seeing the Unbreakable exception.

I am not surprised by what happened in your screenshot (assuming you are building with a shipped compiler and not the one from your feature branch). We try very hard to preserve compatibility between releases, and that means that updating your compiler or language version should not change which method is selected as the entry point. Thus a sync main method should always "win" over an async one, and in the case that we find a sync main method, we don't currently warn if there are multiple async main candidates, for example.

**It feels like the solution might be**: if at least one sync Main method is found, then we should give the new warning on the declarations of all the async Main methods basically saying: "Method 'async Main()' will not be used as an entry point because a synchronous entry point 'void Main(...)' was found."

The formatted message in CSharpResources.resx would probably look like `"Method '{0}' will not be used as an entry point because a synchronous entry point '{1}' was found."`, then pass in the appropriate MethodSymbols to DiagnosticBag.Add.

Reply:
I quite like this approach, thanks for the suggestion @RikkiGibson.